### PR TITLE
i#5843 scheduler: Delay switch to instr after blocking syscall

### PR DESCRIPTION
**Pull request by derekbruening**
Fixes a problem with the implementation of context switching on a blocking syscall: it was being done immediately upon seeing the maybe-blocking marker, which is right after the syscall instruction. This resulted in recording the syscall instruction ordinal as the switch point, when that ordinal is supposed to be exclusive and so needs to be right before the subsequent instruction.

Adds a test that ensures we don't switch until after all the markers after a blocking syscall.

Issue: #5843

### CODE REVIEW

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 1]
```diff
@@ -1210,13 +1210,10 @@ test_synthetic_with_syscalls_multiple()
...
     // blocking syscall.
-    assert(sched_as_string[0] == "BHHHBEEBHHHBCCCBIIICCCFFFIIIFFFAAAGGGDDDAAAGGG");
-    assert(sched_as_string[1] == "EEBEEHHHEEBEBFFFBJJJJJJJJJCCCIIIDDDAAAGGGDDD");
-    // Sanity check that the same input isn't on both cores at once.
-    for (size_t i = 0; i < std::min(sched_as_string[0].size(), sched_as_string[1].size());
-         ++i) {
-        assert(sched_as_string[0][i] != sched_as_string[1][i]);
-    }
+    // Note that the 3rd B is not really on the two cores at the same time as there are
+    // markers and other records that cause them to not actually line up.
```
Comment (abhinav92003) [inline clients/drcachesim/tests/scheduler_unit_tests.cpp:1214]: This was true before this PR too right? So maybe the for-loop check shouldn't've been added at all?

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 2]
```diff
@@ -1282,11 +1279,120 @@ test_synthetic_with_syscalls_single()
...
+{
+    std::cerr << "\n----------------\nTesting blocking syscall precise switch points\n";
+    static constexpr memref_tid_t TID_A = 42;
+    static constexpr memref_tid_t TID_B = 99;
+    static constexpr int SYSNUM = 202;
+    std::vector<trace_entry_t> refs_A = {
+        /* clang-format off */
+        make_thread(TID_A),
+        make_pid(1),
+        make_version(4),
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1307]: Symbolic would be better: switched to TRACE_ENTRY_VERSION.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 3]
```diff
@@ -1282,11 +1279,120 @@ test_synthetic_with_syscalls_single()
...
+        check_ref(refs, idx, TID_A, TRACE_TYPE_MARKER) &&
+        check_ref(refs, idx, TID_A, TRACE_TYPE_MARKER) &&
+        check_ref(refs, idx, TID_A, TRACE_TYPE_MARKER) &&
+        check_ref(refs, idx, TID_A, TRACE_TYPE_MARKER) &&
+        check_ref(refs, idx, TID_A, TRACE_TYPE_MARKER) &&
+        check_ref(refs, idx, TID_A, TRACE_TYPE_MARKER) &&
+        check_ref(refs, idx, TID_A, TRACE_TYPE_MARKER) &&
+        check_ref(refs, idx, TID_A, TRACE_TYPE_MARKER) &&
+        check_ref(refs, idx, TID_A, TRACE_TYPE_MARKER) &&
+        check_ref(refs, idx, TID_A, TRACE_TYPE_MARKER) &&
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1378]: I did that, and also added the marker sub-type to the check call so you can see which one is which.

[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1378]: Yes, not ideal.  Added a comment.  We can discuss further wrt the indirect branch marker discussion offline.

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 4]
```diff
@@ -1739,18 +1739,21 @@ scheduler_tmpl_t<RecordType, ReaderType>::next_record(output_ordinal_t output,
...
             trace_marker_type_t marker_type;
             uintptr_t marker_value;
-            if (record_type_is_marker(record, marker_type, marker_value) &&
-                marker_type == TRACE_MARKER_TYPE_MAYBE_BLOCKING_SYSCALL) {
-                if (!input->processed_blocking_syscall) {
+            if (input->saw_blocking_syscall) {
+                // Wait until we're past all the markers associated with the syscall.
+                // If a signal arrives we would like the kernel xfer markers to go
+                // the next instr: but our recorded format is on instr boundaries.
+                if (record_type_is_instr(record)) {
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:1746]: Separating this out to combine with the ind branch marker placement.  After offline discussion we can file a new issue if we plan to change things.  For now I put a comment here.

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 5]
```diff
@@ -1739,18 +1739,21 @@ scheduler_tmpl_t<RecordType, ReaderType>::next_record(output_ordinal_t output,
...
                     // Assume it will block and we should switch to a different input.
                     need_new_input = true;
                     in_wait_state = true;
-                    // Avoid simply recursing when we're rescheduled.
-                    input->processed_blocking_syscall = true;
+                    input->saw_blocking_syscall = false;
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:1750]: Done.

**Pull request merged**